metagraph/src/graph/representation/masked_graph.cpp

```diff
@@ -278,5 +243,49 @@ bool MaskedDeBruijnGraph::operator==(const DeBruijnGraph &other) const {
     return DeBruijnGraph::operator==(other);
 }
 
+void MaskedDeBruijnGraph::update_mask(const GenerateNodeUpdates &generate_nodes,
+                                      bool in_place, bool async, int memorder) {
+    std::mutex set_mutex;
+    auto *mask_updateable = dynamic_cast<bitmap_dyn*>(kmers_in_graph_.get());
+    if (in_place && mask_updateable) {
+        auto *mask_vector = dynamic_cast<bitmap_vector*>(mask_updateable);
+        if (mask_vector) {
+            auto &mask_sdsl = const_cast<sdsl::bit_vector&>(mask_vector->data());
+            generate_nodes([&](node_index i, bool val) {
+                if (val) {
+                    set_bit(mask_sdsl.data(), i, async, memorder);
+                } else {
+                    unset_bit(mask_sdsl.data(), i, async, memorder);
+                }
+            });
+        } else {
+            if (async) {
+                generate_nodes([&](node_index i, bool val) {
+                    std::lock_guard<std::mutex> lock(set_mutex);
+                    mask_updateable->set(i, val);
+                });
+            } else {
+                generate_nodes([&](node_index i, bool val) {
+                    mask_updateable->set(i, val);
+                });
+            }
+        }
+    } else {
+        if (in_place)
+            logger->warn("Mask not updateable. Making a copy instead");
+
+        sdsl::bit_vector mask_updated(kmers_in_graph_->size(), false);
+        kmers_in_graph_->add_to(&mask_updated);
+        generate_nodes([&](node_index i, bool val) {
+            if (val) {
+                set_bit(mask_updated.data(), i, async, memorder);
+            } else {
+                unset_bit(mask_updated.data(), i, async, memorder);
+            }
+        });
+        kmers_in_graph_ = std::make_unique<bit_vector_stat>(std::move(mask_updated));
+    }
+}
```

Comment:
This looks too complex and I bet it's suboptimal in some cases. So why do we restrict ourselves to this interface?

I think it's better to just return a pointer to the bitmap, so it can be updated externally by the user who may have extra information (for instance, how many bits are going to be updated. If too many, it faster to reinitialize the whole thing).

Alternatively, we can allow only the reinitialization of the whole thing.

Reply:
So simply return a pointer to the bitmap `return &kmers_in_graph_;` and move this code for updating to the caller.

Reply:
The easiest thing will be to just return a non-const reference to the underlying mask and let an external function deal with modifying it.

Reply:
To give motivation for why I'm doing all of this, I want to be able to call `MaskedDeBruijnGraph::call_unitigs` and modify the mask in the callback while avoiding as much RAM overhead as possible.

If the underlying graph is DBGSuccinct, then a copy of the mask is made before traversal starts, so I can modify the mask in the callback without disturbing anything. For any other graph representation, modifying the mask in the callback will break the traversal, so I have to populate a new bit vector first, then swap the masks.

Since this distinction between the graph types is an implementation detail of MaskedDeBruijnGraph, I tried to put all of this logic in `update_mask`.

Reply:
Then return a non-const reference to the `unqiue_ptr` to allow the user to reinitialize the whole thing and change representations.

Reply:
Yeah, I see the need. Just the interface is too restricted and complex.